jobRun.go

```diff
@@ -0,0 +1,284 @@
+package jobsd
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/simpleframeworks/logc"
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+// JobRun .
+type JobRun struct {
+	ID                    int64 `gorm:"primaryKey"`
+	OriginID              int64 `gorm:"index"`
+	Name                  sql.NullString
+	NameActive            sql.NullString `gorm:"unique"`
+	Job                   string
+	JobArgs               Args
+	Delay                 time.Duration
+	RunAt                 time.Time
+	RunCount              int
+	RunLimit              sql.NullInt64
+	RunStartedAt          sql.NullTime
+	RunStartedBy          sql.NullInt64 `gorm:"index"`
+	RunCompletedAt        sql.NullTime
+	RunCompletedError     sql.NullString
+	RetryTimeout          time.Duration
+	RetryTimeoutAt        sql.NullTime `gorm:"index"`
+	RetriesOnErrorCount   int
+	RetriesOnErrorLimit   int
+	RetriesOnTimeoutCount int
+	RetriesOnTimeoutLimit int
+	Schedule              sql.NullString
+	ClosedAt              sql.NullTime
+	ClosedBy              sql.NullInt64 `gorm:"index"`
+	CreatedAt             time.Time     `gorm:"index"`
+	CreatedBy             int64
+}
+
+// insertGet inserts the Job in the DB. If it already exists it retrieves it.
+func (j *JobRun) insertGet(db *gorm.DB) error {
+	if !j.NameActive.Valid {
+		return errors.New("the active name, NameActive, is required")
+	}
+	tx := db.Clauses(clause.OnConflict{DoNothing: true}).Create(j)
+	if tx.Error != nil {
+		return tx.Error
+	}
+	if j.ID == 0 {
+		db.Where("name_active = ?", j.NameActive).First(j)
+	}
+	return tx.Error
+}
+
+// lockStart lock and start the job to run
+func (j *JobRun) lockStart(db *gorm.DB, instanceID int64) (bool, error) {
+	startedAt := time.Now()
+	retryTimeoutAt := startedAt.Add(j.RetryTimeout)
+	tx := db.Model(j).Where("run_started_at IS NULL").Updates(map[string]interface{}{
+		"retry_timeout_at": retryTimeoutAt,
+		"run_started_at":   startedAt,
+		"run_started_by":   instanceID,
+	})
+	if tx.Error != nil {
+		return false, tx.Error
+	}
+	locked := tx.RowsAffected == 1
+	if locked {
+		j.RetryTimeoutAt = sql.NullTime{Valid: true, Time: retryTimeoutAt}
+		j.RunStartedAt = sql.NullTime{Valid: true, Time: startedAt}
+		j.RunStartedBy = sql.NullInt64{Valid: true, Int64: instanceID}
+	}
+	return locked, tx.Error
+}
+
+// complete the job
+func (j *JobRun) complete(db *gorm.DB, instanceID int64, jobRunErr error) error {
+
+	completedErr := sql.NullString{}
+	if jobRunErr != nil {
+		completedErr = sql.NullString{Valid: true, String: jobRunErr.Error()}
+	} else {
+		j.RunCount++
```

Comment:
No. It can only be modified by the worker holding the lock. 

If this operation fails or the server crashes the timeout recovers this and retries the JobRun.